### PR TITLE
Cleaning of singularity fields code

### PR DESCRIPTION
**Pull request by Menshin**
- Cleaned field generator (old) code
- Fixes runtimes on field shutdown

### CODE REVIEW

File: code/modules/power/singularity/field_generator.dm [hunk 1]
```diff
@@ -196,44 +183,37 @@ field_generator power level display
...
+	if(power >= draw)//We have enough power
+		power -= draw
 		return 1
+
 	else//Need more power
-		draw -= src.power
-		src.power = 0
+		draw -= power
+		power = 0
 		for(var/obj/machinery/field/generator/FG in connected_gens)
```
Comment (RemieRichards) [inline code/modules/power/singularity/field_generator.dm]: this should only contain field gens
so it should be an istypeless for loop.

File: code/modules/power/singularity/field_generator.dm [hunk 2]
```diff
@@ -260,82 +240,69 @@ field_generator power level display
...
 			steps -= 1
 			if(!G.active)
 				return 0
 			break
-	if(isnull(G))
-		return
-	T = src.loc
-	for(var/dist = 0, dist < steps, dist += 1) // creates each field tile
+
+		for(var/atom/A in T.contents)
```
Comment (RemieRichards) [inline code/modules/power/singularity/field_generator.dm]: you can only have atoms in contents.
istypeless for loop.

File: code/modules/power/singularity/field_generator.dm [hunk 3]
```diff
@@ -47,51 +53,35 @@ field_generator power level display
...
-			anchored = 1
-			warming_up = 3
-			start_fields()
-			update_icon()
-		Varedit_start = 0
-
-	if(src.active == 2)
+	if(active == ONLINE)
 		calc_power()
 		update_icon()
```
Comment (RemieRichards) [inline code/modules/power/singularity/field_generator.dm]: this was the thing about field gens that needed fixing the most.
this should be an on-demand call to ```update_icon()```, not constantly polled by ```process()```

File: code/modules/power/singularity/field_generator.dm [hunk 4]
```diff
@@ -13,6 +13,15 @@ field_generator power level display
...
 
 #define field_generator_max_power 250
+
+#define UNSECURED 0
+#define SECURED 1
+#define WELDED 2
+
+#define OFFLINE 0
+#define CHARGING 1
+#define ONLINE 2
```
Comment (phil235) [inline code/modules/power/singularity/field_generator.dm]: Are you sure there are no other defines with the same name?

File: code/modules/power/singularity/field_generator.dm [hunk 5]
```diff
@@ -196,44 +183,37 @@ field_generator power level display
...
-		return 1
-
 	update_icon()
-	if(src.power > field_generator_max_power)
-		src.power = field_generator_max_power
+	if(power > field_generator_max_power)
+		power = field_generator_max_power
 
 	var/power_draw = 2
 	for (var/obj/machinery/field/containment/F in fields)
```
Comment (phil235) [inline code/modules/power/singularity/field_generator.dm]: if there are only this type of objects in fields, you could use an istypeless"for" loop.

File: code/modules/power/singularity/field_generator.dm [hunk 6]
```diff
@@ -260,82 +240,69 @@ field_generator power level display
...
-	if(!listcheck)
-		G.connected_gens.Add(src)
+
+	connected_gens |= G
+	G.connected_gens |= src
 
 
 /obj/machinery/field/generator/proc/cleanup()
 	clean_up = 1
 	for (var/obj/machinery/field/containment/F in fields)
```
Comment (phil235) [inline code/modules/power/singularity/field_generator.dm]: istypeless for loop

**Pull request merged**